src/as_symbol_name.rs

```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+impl private::AsSymbolNameSeal for &String {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        self.as_str().symbol_name(function)
+    }
+}
+
+impl private::AsSymbolNameSeal for String {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        let mut data = self.into_bytes();
+        if let Some(position) = crate::util::find_interior_element(&data, 0) {
+            return Err(Error::InteriorZeroElements { position });
+        }
+
+        if data.last() != Some(&0) {
+            data.push(0);
+        }
+
+        function(data.as_ptr().cast())
+    }
+}
+
+impl private::AsSymbolNameSeal for &CStr {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        function(self.as_ptr())
+    }
+}
+
+impl private::AsSymbolNameSeal for &CString {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        function(self.as_ptr())
+    }
+}
+
+impl private::AsSymbolNameSeal for CString {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        function(self.as_ptr())
+    }
+}
+
+impl private::AsSymbolNameSeal for &[u8] {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        if let Some(position) = crate::util::find_interior_element(self, 0) {
+            return Err(Error::InteriorZeroElements { position });
+        }
+
+        if self.last() != Some(&0) {
+            let copy = crate::util::copy_and_push(self, 0);
+            return function(copy.as_ptr().cast());
+        }
+
+        function(self.as_ptr().cast())
+    }
+}
+
+impl<const N: usize> private::AsSymbolNameSeal for [u8; N] {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        self.as_slice().symbol_name(function)
+    }
+}
+
+impl<const N: usize> private::AsSymbolNameSeal for &[u8; N] {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        self.as_slice().symbol_name(function)
+    }
+}
+
+/// This implementation requires that the buffer contains valid data to call [`String::from_utf16`].
+impl private::AsSymbolNameSeal for &[u16] {
+    fn symbol_name<R>(
+        self,
+        function: impl FnOnce(*const core::ffi::c_char) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        let string = String::from_utf16(self)?;
+        string.symbol_name(function)
+    }
+}
```

Comment:
I still don't see the point of this implementation, really. It only hides from the end user the cost of passing in this sort of type.